### PR TITLE
Refactoring for Windows support

### PR DESCRIPTION
**Pull request by kelunik**
This is WIP, it currently segfaults.

### CODE REVIEW

File: lib/Internal/Posix/Runner.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            Loop::cancel($handle->extraDataPipeStartWatcher);
+            $handle->extraDataPipeStartWatcher = null;
+        }
+
+        if (!\proc_terminate($handle->proc, 9)) { // Forcefully kill the process using SIGKILL.
+            throw new ProcessException("Terminating process failed");
+        }
+
+        $handle->status = ProcessStatus::ENDED;
+        $handle->joinDeferred->fail(new ProcessException("The process was killed"));
```
Comment (bwoebi) [inline lib/Internal/Posix/Runner.php]: I don't think that's what should happen here - we should return all the remaining data on the pipe until now instead.

File: lib/Internal/Posix/Handle.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+final class Handle extends ProcessHandle {
+    public function __construct() {
+        $this->pidDeferred = new Deferred;
+        $this->joinDeferred = new Deferred;
+        $this->originalParentPid = \getmypid();
+    }
+
+    /** @var Deferred */
+    public $joinDeferred;
```
Comment (bwoebi) [inline lib/Internal/Posix/Handle.php]: this really is an `exitCodeDeferred`, right? `joinDeferred` does not tell me much about what will be returned from it.

Comment (kelunik) [reply lib/Internal/Posix/Handle.php]: The method returning the exit code is called `join` as well.

Comment (bwoebi) [reply lib/Internal/Posix/Handle.php]: I am aware, I just found it confusing while reading the code.

**Comment by trowski** [comment]
@kelunik What's the status here? Did we want to move forward with this or look at the version where `start()` returns a promise?

**Comment by kelunik** [comment]
I wanted to rewrite parts of it, but that can happen later. It works the way it currently is. I'm mostly waiting for you having more time again, so we can get out a release for `amphp/parallel` at the same time, so people don't run into conflicts.

**Comment by kelunik** [comment]
@trowski Did you merge this manually and forgot to close?

**Comment by bwoebi** [comment]
Looks like he rebased your commits. So yes, closing.

**Pull request closed without merging**